### PR TITLE
feat(tf/pt): add/refact lammps support for spin model

### PR DESCRIPTION
**Pull request by iProzd**
## Summary by CodeRabbit

- **New Features**
	- Introduced spin handling capabilities in various models, enhancing energy and force computations.
	- Added support for new parameters in compute methods to accommodate ghost atoms and neighbor lists.
	- Implemented new classes and methods for managing spin interactions, including `DeepSpin`, `DeepSpinModelDevi`, and their TensorFlow and PyTorch counterparts.
	- Added a new pair style for LAMMPS simulations: `PairDeepSpin`.
	- Enhanced the `DeepPot` framework with spin support, including new API structures and functions.
	- Introduced a new function for writing LAMMPS data that includes spin information.
	- Added comprehensive functionality for error handling and tensor management in `DeepSpinPT` and `DeepSpinTF`.
	- Established a new testing framework for the `DeepSpin` class to validate its methods under various configurations.

- **Bug Fixes**
	- Improved error handling in TensorFlow and PyTorch integrations to ensure robustness during model execution.

- **Tests**
	- Established comprehensive unit tests for the `DeepSpin` class, validating spin-related computations and ensuring accuracy in energy and force calculations.
	- Introduced additional tests for the `DeepSpin` class, focusing on the `compute` method and verifying outputs against expected values.
	- Added unit tests for the `DeepSpinPT` class to validate its functionality.
	- Enhanced testing framework for LAMMPS spin simulations, ensuring correctness across various scenarios.
	- Developed tests for the `DeepSpin` class to validate its behavior with and without periodic boundary conditions.

- **Documentation**
	- Enhanced documentation for new methods and classes to clarify usage and functionality related to spin interactions.

### CODE REVIEW

File: source/api_c/include/c_api.h
Comment (njzjz) [inline source/api_c/include/c_api.h:1]: Please bump the C API version:
https://github.com/deepmodeling/deepmd-kit/blob/3466e34b5dacafb47c071d3affde9747f52c503b/source/api_c/include/c_api.h#L12-L15

File: source/lmp/pair_deepspin.cpp [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+      (numb_models > 1 && (out_freq == 0 || update->ntimestep % out_freq != 0));
+  multi_models_mod_devi =
+      (numb_models > 1 && (out_freq > 0 && update->ntimestep % out_freq == 0));
+  if (do_ghost) {
+    deepmd_compat::InputNlist lmp_list(
+        list->inum, list->ilist, list->numneigh, list->firstneigh,
+        commdata_->nswap, commdata_->sendnum, commdata_->recvnum,
+        commdata_->firstrecv, commdata_->sendlist, commdata_->sendproc,
+        commdata_->recvproc, &world);
+    deepmd_compat::InputNlist extend_lmp_list;
```
Comment (hztttt) [inline source/lmp/pair_deepspin.cpp:211]: we don't need 'extend_lmp_list' here.

File: source/lmp/tests/run_mpi_pair_deepmd_spin.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+lammps.neigh_modify("every 10 delay 0 check no")
+lammps.read_data(data_file)
+lammps.mass("1 58")
+lammps.mass("2 16")
+lammps.timestep(0.0005)
+lammps.fix("1 all nve")
+
+relative = 1.0
+lammps.pair_style(
+    f"deepmd {pb_file} {pb_file2} out_file {md_file} out_freq 1 atomic relative {relative}"
```
Comment (hztttt) [inline source/lmp/tests/run_mpi_pair_deepmd_spin.py:57]: need to change 'deepmd' into 'deepspin'.

File: source/lmp/plugin/deepmdplugin.cpp [hunk 3]
```diff
@@ -41,6 +43,15 @@ extern "C" void lammpsplugin_init(void *lmp, void *handle, void *regfunc) {
   plugin.handle = handle;
   (*register_plugin)(&plugin, lmp);
 
+  plugin.version = LAMMPS_VERSION;
+  plugin.style = "pair";
+  plugin.name = "deepspin";
+  plugin.info = "deepspin pair style " STR_GIT_SUMM;
+  plugin.author = "Duo Zhang";
```
Comment (njzjz) [inline source/lmp/plugin/deepmdplugin.cpp:50]: This line may need to be reset after line 55 - note that we never re-initialize `plugin`, so the attributes in the previous plugin are kept.

File: source/lmp/pair_deepmd.cpp [hunk 4]
```diff
@@ -563,60 +200,26 @@ void PairDeepMD::compute(int eflag, int vflag) {
         commdata_->nswap, commdata_->sendnum, commdata_->recvnum,
         commdata_->firstrecv, commdata_->sendlist, commdata_->sendproc,
         commdata_->recvproc, &world);
-    lmp_list.set_mask(NEIGHMASK);
```
Comment (njzjz) [inline source/lmp/pair_deepmd.cpp:566]: #4269 should not be reverted, and you may also need to apply it to your new code.

**Review by njzjz — CHANGES_REQUESTED** [review]

**Comment by iProzd** [comment]
The conversations are not readable. I'll open up another clean PR for review.
